### PR TITLE
Fix example code from How to write a type class derived method using macros

### LINKED ISSUE #10381: Code from docs does not compile (typeclass derivation with macros)
Hi! I was trying to learn how to derive typeclass instances with macros but the code from [How to write a type class `derived` method using macros](https://dotty.epfl.ch/docs/reference/contextual/derivation-macro.html) does not compile with 3.0.0-M1.

## Minimized example

```Scala
import scala.deriving._
import scala.quoted._
import scala.quoted.matching._

trait Eq[T] {
  def eqv(x: T, y: T): Boolean
}

object Eq {
  given Eq[String] {
    def eqv(x: String, y: String) = x == y
  }

  given Eq[Int] {
    def eqv(x: Int, y: Int) = x == y
  }

  def eqProduct[T](body: (T, T) => Boolean): Eq[T] =
    new Eq[T] {
      def eqv(x: T, y: T): Boolean = body(x, y)
    }

  def eqSum[T](body: (T, T) => Boolean): Eq[T] =
    new Eq[T] {
      def eqv(x: T, y: T): Boolean = body(x, y)
    }

  def summonAll[T](t: Type[T])(using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
    case '[String *: $tpes] => '{ summon[Eq[String]] }  :: summonAll(tpes)
    case '[Int *: $tpes]    => '{ summon[Eq[Int]] }     :: summonAll(tpes)
    case '[$tpe *: $tpes]   => derived(using tpe, qctx) :: summonAll(tpes)
    case '[EmptyTuple] => Nil
  }

  given derived[T: Type](using qctx: QuoteContext) as Expr[Eq[T]] = {
    import qctx.reflect._

    val ev: Expr[Mirror.Of[T]] = Expr.summon(using '[Mirror.Of[T]]).get

    ev match {
      case '{ $m: Mirror.ProductOf[T] { type MirroredElemTypes = $elementTypes }} =>
        val elemInstances = summonAll(elementTypes)
        val eqProductBody: (Expr[T], Expr[T]) => Expr[Boolean] = (x, y) => {
          elemInstances.zipWithIndex.foldLeft(Expr(true: Boolean)) {
            case (acc, (elem, index)) =>
              val e1 = '{$x.asInstanceOf[Product].productElement(${Expr(index)})}
              val e2 = '{$y.asInstanceOf[Product].productElement(${Expr(index)})}

              '{ $acc && $elem.asInstanceOf[Eq[Any]].eqv($e1, $e2) }
          }
        }
        '{
          eqProduct((x: T, y: T) => ${eqProductBody('x, 'y)})
        }

      case '{ $m: Mirror.SumOf[T] { type MirroredElemTypes = $elementTypes }} =>
        val elemInstances = summonAll(elementTypes)
        val eqSumBody: (Expr[T], Expr[T]) => Expr[Boolean] = (x, y) => {
          val ordx = '{ $m.ordinal($x) }
          val ordy = '{ $m.ordinal($y) }

          val elements = Expr.ofList(elemInstances)
          '{
              $ordx == $ordy && $elements($ordx).asInstanceOf[Eq[Any]].eqv($x, $y)
          }
        }

        '{
          eqSum((x: T, y: T) => ${eqSumBody('x, 'y)})
        }
    }
  }
}

object Macro3 {
  extension [T](x: =>T)
    inline def === (y: =>T)(using eq: Eq[T]): Boolean = eq.eqv(x, y)

  implicit inline def eqGen[T]: Eq[T] = ${ Eq.derived[T] }
}
```

## Output

```scala
[error] -- [E006] Not Found Error: /Users/pierre/GIT/dotty-example-project/src/main/scala/Test.scala:29:71 
[error] 29 |      case '[String *: $tpes] => '{ summon[Eq[String]] }  :: summonAll(tpes)
[error]    |                                                                       ^^^^
[error]    |                                                         Not found: tpes
[error] -- [E006] Not Found Error: /Users/pierre/GIT/dotty-example-project/src/main/scala/Test.scala:30:71 
[error] 30 |      case '[Int *: $tpes]    => '{ summon[Eq[Int]] }     :: summonAll(tpes)
[error]    |                                                                       ^^^^
[error]    |                                                         Not found: tpes
[error] -- [E006] Not Found Error: /Users/pierre/GIT/dotty-example-project/src/main/scala/Test.scala:31:71 
[error] 31 |      case '[$tpe *: $tpes]   => derived(using tpe, qctx) :: summonAll(tpes)
[error]    |                                                                       ^^^^
[error]    |                                                         Not found: tpes
[error] -- [E006] Not Found Error: /Users/pierre/GIT/dotty-example-project/src/main/scala/Test.scala:31:47 
[error] 31 |      case '[$tpe *: $tpes]   => derived(using tpe, qctx) :: summonAll(tpes)
[error]    |                                               ^^^
[error]    |                                               Not found: tpe
[warn] -- Warning: /Users/pierre/GIT/dotty-example-project/src/main/scala/Test.scala:38:53 
[warn] 38 |      val ev: Expr[Mirror.Of[T]] = Expr.summon(using '[Mirror.Of[T]]).get
[warn]    |                                                     ^^^^^^^^^^^^^^^
[warn]    |Consider using canonical type constructor scala.quoted.Type[Mirror.Of[T]] instead
[error] -- [E006] Not Found Error: /Users/pierre/GIT/dotty-example-project/src/main/scala/Test.scala:42:40 
[error] 42 |          val elemInstances = summonAll(elementTypes)
[error]    |                                        ^^^^^^^^^^^^
[error]    |                                        Not found: elementTypes
[error] -- [E006] Not Found Error: /Users/pierre/GIT/dotty-example-project/src/main/scala/Test.scala:57:40 
[error] 57 |          val elemInstances = summonAll(elementTypes)
[error]    |                                        ^^^^^^^^^^^^
[error]    |                                        Not found: elementTypes
[warn] one warning found
[error] 6 errors found
```

### PR DESCRIPTION
**Pull request by ghostdogpr**
Closes #10381

### CODE REVIEW

**Comment by ghostdogpr** [comment]
(I just signed the CLA)

**Review by smarter — APPROVED** [review]
Thanks! @nicolasstucki Can you double-check that this is still a good example for the documentation?

File: docs/docs/reference/contextual/derivation-macro.md [hunk 1]
```diff
@@ -84,17 +84,17 @@ Instead we extract the tuple-type for element types using pattern matching over
...
 
 The implementation of `summonAll` as a macro can be show below assuming that we
 have the given instances for our primitive types:
 
 ```scala
-  def summonAll[T](t: Type[T])(using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
-    case '[String *: $tpes] => '{ summon[Eq[String]] }  :: summonAll(tpes)
-    case '[Int *: $tpes]    => '{ summon[Eq[Int]] }     :: summonAll(tpes)
-    case '[$tpe *: $tpes]   => derived(using tpe, qctx) :: summonAll(tpes)
+  def summonAll[T](using t: Type[T])(using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
```
Comment (nicolasstucki) [inline docs/docs/reference/contextual/derivation-macro.md:94]:
```suggestion
  def summonAll[T: Type](using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
```

File: docs/docs/reference/contextual/derivation-macro.md [hunk 2]
```diff
@@ -169,21 +169,21 @@ object Eq {
       def eqv(x: T, y: T): Boolean = body(x, y)
     }
 
-  def summonAll[T](t: Type[T])(using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
-    case '[String *: $tpes] => '{ summon[Eq[String]] }  :: summonAll(tpes)
-    case '[Int *: $tpes]    => '{ summon[Eq[Int]] }     :: summonAll(tpes)
-    case '[$tpe *: $tpes]   => derived(using tpe, qctx) :: summonAll(tpes)
+  def summonAll[T](using t: Type[T])(using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
```
Comment (nicolasstucki) [inline docs/docs/reference/contextual/derivation-macro.md:172]:
```suggestion
  def summonAll[T: Type](using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
```

File: docs/docs/reference/contextual/derivation-macro.md [hunk 3]
```diff
@@ -217,8 +217,7 @@ object Eq {
 }
 
 object Macro3 {
-  extension [T](x: =>T)
-    inline def === (y: =>T)(using eq: Eq[T]): Boolean = eq.eqv(x, y)
+  extension [T](inline x: T) inline def === (inline y: T)(using eq: Eq[T]): Boolean = eq.eqv(x, y)
```
Comment (nicolasstucki) [inline docs/docs/reference/contextual/derivation-macro.md:220]:
```suggestion
  extension [T](inline x: T) 
    inline def === (inline y: T)(using eq: Eq[T]): Boolean = eq.eqv(x, y)
```

**Review by nicolasstucki — APPROVED** [review]
LGTM. I just added a couple of extra simplification.

File: docs/docs/reference/contextual/derivation-macro.md [hunk 4]
```diff
@@ -44,11 +44,11 @@ from the signature. The body of the `derived` method is shown below:
 given derived[T: Type](using qctx: QuoteContext) as Expr[Eq[T]] = {
   import qctx.reflect._
 
-  val ev: Expr[Mirror.Of[T]] = Expr.summon(using '[Mirror.Of[T]]).get
+  val ev: Expr[Mirror.Of[T]] = Expr.summon(using Type.of[Mirror.Of[T]]).get
```
Comment (nicolasstucki) [inline docs/docs/reference/contextual/derivation-macro.md:47]:
```suggestion
  val ev: Expr[Mirror.Of[T]] = Expr.summon[Mirror.Of[T]].get
```

File: docs/docs/reference/contextual/derivation-macro.md [hunk 5]
```diff
@@ -169,21 +169,21 @@ object Eq {
...
+    case '[Int *: tpes]    => '{ summon[Eq[Int]] }     :: summonAll[tpes]
+    case '[tpe *: tpes]   => derived[tpe] :: summonAll[tpes]
     case '[EmptyTuple] => Nil
   }
 
   given derived[T: Type](using qctx: QuoteContext) as Expr[Eq[T]] = {
     import qctx.reflect._
 
-    val ev: Expr[Mirror.Of[T]] = Expr.summon(using '[Mirror.Of[T]]).get
+    val ev: Expr[Mirror.Of[T]] = Expr.summon(using Type.of[Mirror.Of[T]]).get
```
Comment (nicolasstucki) [inline docs/docs/reference/contextual/derivation-macro.md:182]:
```suggestion
    val ev: Expr[Mirror.Of[T]] = Expr.summon[Mirror.Of[T]].get
```

**Comment by nicolasstucki** [comment]
Thank you @ghostdogpr

File: docs/docs/reference/contextual/derivation-macro.md [hunk 6]
```diff
@@ -84,17 +84,17 @@ Instead we extract the tuple-type for element types using pattern matching over
...
 
 The implementation of `summonAll` as a macro can be show below assuming that we
 have the given instances for our primitive types:
 
 ```scala
-  def summonAll[T](t: Type[T])(using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
-    case '[String *: $tpes] => '{ summon[Eq[String]] }  :: summonAll(tpes)
-    case '[Int *: $tpes]    => '{ summon[Eq[Int]] }     :: summonAll(tpes)
-    case '[$tpe *: $tpes]   => derived(using tpe, qctx) :: summonAll(tpes)
+  def summonAll[T: Type](using qctx: QuoteContext): List[Expr[Eq[_]]] = t match {
```
Comment (smarter) [inline docs/docs/reference/contextual/derivation-macro.md:94]: @nicolasstucki Does this actually compile? I don't see where `t` is defined here. Also if we update the doc we should keep the test case synchronized: https://github.com/lampepfl/dotty/blob/master/tests/run-macros/i8007/Macro_3.scala

Comment (nicolasstucki) [reply docs/docs/reference/contextual/derivation-macro.md:94]: I missed that. I will fix it.

Comment (nicolasstucki) [reply docs/docs/reference/contextual/derivation-macro.md:94]: Fix in #10395

**Pull request merged**